solr/core/src/java/org/apache/solr/core/ConfigSetService.java

```diff
@@ -391,6 +402,21 @@ protected NamedList<Object> loadConfigSetFlags(SolrResourceLoader loader) throws
    */
   protected abstract SolrResourceLoader createCoreResourceLoader(CoreDescriptor cd);
 
+  /**
+   * Create a SolrResourceLoader for a core with the provided configSetName.
+   *
+   * <p>By default, this will just call {@link
+   * ConfigSetService#createConfigSetService(CoreContainer)}. Child implementation might override
+   * this to make use of the configSetName directly
+   *
+   * @param cd the core's CoreDescriptor
+   * @param configSetName an optional config set name
+   * @return a SolrResourceLoader
+   */
+  protected SolrResourceLoader createCoreResourceLoader(CoreDescriptor cd, String configSetName) {
```

Comment:
Since the configSetName is also on the CoreDescriptor, isn't this API needless?  Same elsewhere (loadConfigSet)

Reply:
Yes the choice I took was problematic! 😅 

The ultimate goal is to minimize code change and yet avoid calling [this part of the code](https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java#L77-L100) for the synthetic core, so we wouldn't be registering against ZK. 

Since I saw the quoted code above "reads the config name from cluster state and set it against the CoreDescriptor", so I thought maybe a new "explicit" configset can indicate that we can "skip" those logic for `ZkConfigSetService`.

But it didn't turn out well if I confused other devs 😞 ...

Adding any flags/override method in `ConfigSetService` is likely not great since other non ZK implementation shouldn't really care. So im proposing to add a new class `SythethticCoreDescriptor` which extends `CoreDescriptor`, and that `ZkConfigSetService#createCoreResourceLoader` could do a special case (`instanceof  SythethticCoreDescriptor`, not great...but at least it's all contained in Zk related class) to bypass the zk logic.

Going to commit and push the change, please do lemme know if there are better alternatives!

Reply:
> The ultimate goal is to minimize code change and yet avoid calling [this part of the code](https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java#L77-L100) for the synthetic core, so we wouldn't be registering against ZK.

Why?  It appears it'd have no ultimate effect -- it ought to resolve the same config for the same collection.  Oh, is it the same collection for this feature?

I don't love subclassing CoreDescriptor to be honest; you could just add a generic property in the core descriptor if you needed to flag one like "synthetic".  Nonetheless overall I'm unsure why loading the ConfigSet needs to be different here.

Also if for some reason we felt it useful, we could change ZkConfigSetServer.loadConfigSet such that an existing configSet name in the CD will be used, and thus avoid resolving that.

Reply:
@dsmiley thank you for the review again!

>Why? It appears it'd have no ultimate effect -- it ought to resolve the same config for the same collection. Oh, is it the same collection for this feature?

We want to avoid registering the synthetic collection with ZK, since it creates various issues for our various tooling which watches the ZK for collection changes. We have to do various checks (and it's very easy to forget for new code) to skip the synthetic collection, which is not ideal. 

>I don't love subclassing CoreDescriptor to be honest; you could just add a generic property in the core descriptor if you needed to flag one like "synthetic". Nonetheless overall I'm unsure why loading the ConfigSet needs to be different here.

>Also if for some reason we felt it useful, we could change ZkConfigSetServer.loadConfigSet such that an existing configSet name in the CD will be used, and thus avoid resolving that.

Got it! If it's more preferable to add property to `CoreDescriptor` instead of subclassing, we could totally do that! I thought we want to minimize changes to existing class hence didn't touch `CoreDescriptor` 😓 

Currently `ZkConfigSetService#createCoreResourceLoader` does 3 things as far as I can tell:
1. Auto create collection in ZK if it's not already here
2. Set the configset read from ZK (off the collection from cluster state) to CD. Ignore the pre-existing on in CD
3. Create a `ZkSolrResourceLoader` with access to the resources under the configset in ZK
 
Honestly, I couldn't quite tell the purposes of 1 and 2 (probably not the use cases I'm aware of, I am sure you will be more knowledgeable!) . For synthetic core, we want to avoid both since the for 1., we do not want to register the synthetic collection to ZK and for 2., since the collection is not in ZK it would fail loading such collection from cluster stateu. 

We do want 3 tho as the synthetic core is there to serve resources from the ZK configset.


Please let me know if it all makes sense? If so, I will proceed to add property to `CoreDescriptor` 😊

Reply:
We should remove "1. Auto create collection in ZK if it's not already here"; this refers to legacy behavior.  Or gate this with "bootstrap_conf" system property being specified or something like that.  Heck, I wonder if this chunk of code actually works.... like if you did a little experiment locally, could you get the collection to appear merely by having a core descriptor for a non-existent collection?  I wonder if it's even tested (easy to check by trial and error).

Hopefully there will be no need to designate this CoreDescriptor as special.

There's legacy stuff you are seeing in Solr.  We don't need to support it forever and ever!  Certainly not if it's getting in our way and causing us to contemplate adding otherwise needless complexity.

Reply:
So i tried removing 
```
String configSetName =
          zkController.getClusterState().getCollection(cd.getCollectionName()).getConfigName();
      cd.setConfigSet(configSetName);
```

Then the core loading starts failing. With some investigation:
1. `CoreDescriptor` instantiation does not set the `configSet` in the `coreProperties` via both admin create path and startup core discovery path. Those paths would set `collection.configName` instead. 
2. Therefore without setting it as a part of `ZkConfigSetService#createCoreResourceLoader`, `cd.getConfigSet` would give null, and could trigger error when calling `new ZkSolrResourceLoader(
        cd.getInstanceDir(), cd.getConfigSet(), parentLoader.getClassLoader(), zkController)` (or anywhere else relies on `CoreDescriptor#getgetConfigSet`

I'm not 100% sure if calling `cd.setConfigSet(configSetName)` is a good thing to do within `createCoreResourceLoader` since it appears to be an unexpected side effect. However, to focus on synthetic core alone. I pushed a commit to do a null check with a side note
```
    // Currently, cd.getConfigSet() is always null. Except that it's explicitly set by
    // {@link org.apache.solr.core.SyntheticSolrCore}.
    // Should we consider setting it for all cores as a part of CoreDescriptor creation/loading
    // process?
    if (cd.getConfigSet() == null) {
      String configSetName =
          zkController.getClusterState().getCollection(cd.getCollectionName()).getConfigName();
      cd.setConfigSet(configSetName);
    }
```


Any thoughts? @dsmiley  😊

Reply:
I could understand that it's questionable for the ConfigSetService to be responsible for resolving what the proper named config even ought to be.  But it works and there's not an obvious other location, and it empowers this ConfigSetService abstraction to do interesting things if needed, so I suggest let it be.

Reply:
Thanks so much for doing that manual check and being able to remove old legacy code :-)